### PR TITLE
Download a query result and hide legends if user is not logged in

### PR DESCRIPTION
**Pull request by jorgeatgu**
Closes https://github.com/PopulateTools/issues/issues/1015
https://github.com/PopulateTools/issues/issues/959

## :v: What does this PR do?

Now you can download a file for any query in format CSV, JSON and XLSX.
Hide legends(your queries, favorites, your visualizations) if the user is not logged in.

## :mag: How should this be manually tested?

Staging

**Download files:**
download 2020-05-07 17_05_56

**Hide legends:**
hide-legends 2020-05-07 17_09_45

### CODE REVIEW

File: app/javascript/gobierto_data/lib/factories/download.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import axios from 'axios';
+
+// Data-endpoint factory to get/post/put/delete API data
```
Comment (Crashillo) [inline app/javascript/gobierto_data/lib/factories/download.js]: Copy paste

**Comment by ferblape** [comment]
Careful with the conflicts now we have merged the autocomplete PR

File: app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue [hunk 2]
```diff
@@ -52,32 +73,68 @@ export default {
...
+
+      const {
+        params: {
+          id: titleFile
+        }
+      } = this.$route
+
+      this.titleFile = titleFile
+
+      this.arrayFormatsQuery = [
```
Comment (ferblape) [inline app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue]: Let's make the list of formats dependant on the values that the API returns. Formats are one of the attributes returned in the meta of a dataset:

Screenshot 2020-05-07 at 17 32 58

**Review by ferblape — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue (hunk 2 above)
Comment (furilo) [reply app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue]: We already talked about this in the past, I think...

File: app/javascript/gobierto_data/webapp/components/Dataset.vue [hunk 3]
```diff
@@ -228,6 +233,12 @@ export default {
       queriesPromises.push(this.getPrivateQueries(userId));
     }
 
+    if (!userId || userId.length === 0) {
+      this.isUserLogged = false
+    } else {
+      this.isUserLogged = true
+    }
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/Dataset.vue]: Why do you do a negative clause? I think it's simpler:

```suggestion
      this.isUserLogged = userId && userId.length
```
FYI
https://scotch.io/bar-talk/5-tips-to-write-better-conditionals-in-javascript

Comment (Crashillo) [reply app/javascript/gobierto_data/webapp/components/Dataset.vue]: I've done even a performance test https://jsben.ch/AAi5F, the difference are not so remarkable but it wins positive clauses

File: app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue [hunk 4]
```diff
@@ -13,37 +13,58 @@
...
-              class="gobierto-data-btn-download-data-modal-element"
+            <template
+              v-for="(item, key, index) in arrayFormats"
             >
-              {{ key }}
-            </a>
-          </template>
-        </div>
+              <a
+                :key="index"
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue]: Don't use array indexes as v-for `:key`. It's discouraged. Use the key itself.

FYI
https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318

File: app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue [hunk 5]
```diff
@@ -13,37 +13,58 @@
...
+        <template v-else>
+          <div
+            v-show="!isHidden"
+            class="gobierto-data-btn-download-data-modal"
+          >
+            <template
+              v-for="({ url, name, label }, index) in arrayFormatsQuery"
+            >
+              <a
+                :key="index"
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue]: Same as above

File: app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue [hunk 6]
```diff
@@ -13,37 +13,58 @@
...
 import { baseUrl, CommonsMixin } from "./../../../lib/commons.js";
+import { DownloadFilesFactoryMixin } from "./../../../lib/factories/download";
+
 export default {
   name: 'DownloadButton',
   components: {
     Button
   },
-  mixins: [CommonsMixin],
+  mixins: [CommonsMixin, DownloadFilesFactoryMixin],
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue:67]: Why did we need the CommonsMixin here?

File: app/javascript/gobierto_data/webapp/components/sets/SummaryTab.vue [hunk 7]
```diff
@@ -61,6 +62,10 @@ export default {
       required: true,
       default: () => [],
     },
+    isUserLogged: {
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/SummaryTab.vue:65]: Do you see what annoying is the children props propagation, huh? :D That's why I tell you about the `v-slot`, containers and so...

File: app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorCode.vue [hunk 8]
```diff
@@ -107,32 +110,55 @@ export default {
...
+        }
+      } = editor
+
+      // Show autocomplete only when a letter key is pressed
+      if (!isEditorActive && e.keyCode > 64 && e.keyCode < 91) {
+        editor.showHint({ completeSingle: false })
+      }
+
+      this.$root.$emit('enableSavedButton')
+      this.$root.$emit('enabledForkPrompt')
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/sets/data/SQLEditorCode.vue:133]: Explain the events

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_data/webapp/components/Dataset.vue (hunk 3 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/Dataset.vue]: we need to check if user is not logged, in this case the API return an empty string.

`this.isUserLogged = userId && userId.length`
return a `number`, we need a `Boolean`.

File: app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue (hunk 4 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue]: Thanks for the tip

File: app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue (hunk 6 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue:67]: Yes. We need `clickoutside` [to close the modal.](https://github.com/PopulateTools/gobierto/blob/640ce135e5e0c19cbe31b1def6178fa40770d491/app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue#L4)

File: app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue (hunk 2 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue]: Done @ferblape

File: app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue [hunk 9]
```diff
@@ -52,32 +73,58 @@ export default {
...
       this.isHidden = true
     },
-    updateCode(sqlQuery) {
-      // TODO: this method do nothing
-      const code = sqlQuery
-      const endPointSQL = `${baseUrl}/data.csv?sql=`
-      this.sqlfileCSV = `${endPointSQL}${code}&csv_separator=semicolon`
-      this.sqlfileXLSX = `${endPointSQL}${code}`
-      this.sqlfileJSON = `${endPointSQL}${code}`
+    createObjectFormatsQuery(sql) {
```
Comment (ferblape) [inline app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue]: 👍 much better

File: app/javascript/gobierto_data/webapp/components/Dataset.vue (hunk 3 above)
Comment (ferblape) [reply app/javascript/gobierto_data/webapp/components/Dataset.vue]: Hahah I love Javascript, it's almost a language.

Why not `this.isUserLogged = (!userId || userId.length === 0)` ? Just asking

Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/Dataset.vue]: `this.isUserLogged = (!userId || userId.length === 0)` Return a `Number` if user is logged in, and return `Boolean` if user is not logged

`this.isUserLogged = (userId.length > 0)` always return a `Boolean`

**Review by ferblape — APPROVED** [review]

File: app/javascript/gobierto_data/webapp/components/Dataset.vue (hunk 3 above)
Comment (Crashillo) [reply app/javascript/gobierto_data/webapp/components/Dataset.vue]: If you do something like this `!!(!userId || userId.length === 0)` or `!!(userId && userId.length)` you enforce the Boolean result

https://medium.com/better-programming/javascript-bang-bang-i-shot-you-down-use-of-double-bangs-in-javascript-7c9d94446054

File: app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue [hunk 10]
```diff
@@ -52,32 +73,58 @@ export default {
...
+      } = this.$route
+
+      this.titleFile = titleFile
+
+      for (let i = 0; i < keyFomarts.length; i++) {
+        this.arrayFormatsQuery[i] = {
+          label: keyFomarts[i],
+          url: `${baseUrl}/data.${keyFomarts[i]}?sql=${formatSQL}`,
+          name: `${titleFile}_${date}.${keyFomarts[i]}`
+        };
```
Comment (Crashillo) [inline app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue]: Regarding this file: https://github.com/PopulateTools/gobierto/pull/3038/files#diff-7c5f83f1c0e909e633c79bc133cadc39

Wouldn't it be better approach to move all the` baseUrl`, `keyFomarts` (btw, fix the typo), to that factory file? From this file you only pass the parameters, the url is built in the factory

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue (hunk 10 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_data/webapp/components/commons/DownloadButton.vue]: You're right. Done! ;)

File: app/javascript/gobierto_data/lib/factories/download.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+export const DownloadFilesFactoryMixin = {
+  methods: {
+    createObjectFormatsQuery(sql) {
+
+      //Get the formats from API
+      const keyFormats = Object.keys(this.arrayFormats)
+      //Convert all linebreaks from any SO(Windows: \r\n Linux: \n Older Macs: \r) to spaces.
+      const formatSQL = sql.replace(/[\r\n]+/gm, " ");
+      let datetime = new Date();
+      const date = `${datetime.getDate()}_${(datetime.getMonth() + 1)}_${datetime.getFullYear()}`;
```
Comment (Crashillo) [inline app/javascript/gobierto_data/lib/factories/download.js:18]: Try moment

**Review by Crashillo — APPROVED** [review]

**Review by ferblape — CHANGES_REQUESTED** [review]
This query returns an error downloading the results: https://getafe.gobify.net/datos/censo-secciones/q/123

It seems an error scaping the query in the URL

**Comment by jorgeatgu** [comment]
> This query returns an error downloading the results: https://getafe.gobify.net/datos/censo-secciones/q/123
> 
> It seems an error scaping the query in the URL

Finally, I used the same method we used in the editor to parse the query and turn it into a URL `new URLSearchParams()` parse query with `encodeURI`, `encodeURIComponent` in this case is a headache

**Review by ferblape — APPROVED** [review]

**Pull request merged**